DevTools/HackStudio/Git/DiffViewer.cpp

```diff
@@ -0,0 +1,253 @@
+            if (line_type == LineType::Diff) {
+                painter.fill_rect(
+                    outline,
+                    green_background());
+            }
+            if (line_type == LineType::Missing) {
+                painter.fill_rect(
+                    outline,
+                    gray_background());
+            }
+        }
+    }
+}
+
+size_t DiffViewer::line_height() const
+{
+    return font().glyph_height() + 4;
+}
+
+Gfx::IntRect DiffViewer::separator_rect() const
+{
+    return Gfx::IntRect { frame_inner_rect().width() / 2 - 2,
+        0,
+        4,
+        frame_inner_rect().height() };
+}
+
+void DiffViewer::set_content(const String& original, const String& diff)
+{
+    m_original_lines = split_to_lines(original);
+    m_hunks = Diff::parse_hunks(diff);
+
+#ifdef DEBUG_DIFF
+    for (size_t i = 0; i < m_original_lines.size(); ++i) {
+        dbg() << i << ":" << m_original_lines[i];
+    }
+#endif
+}
+
+DiffViewer::DiffViewer()
+{
+    setup_properties();
+}
+
+DiffViewer::DiffViewer(const String& original, const String& diff)
+    : m_original_lines(split_to_lines(original))
+    , m_hunks(Diff::parse_hunks(diff))
+{
+    setup_properties();
+}
+
+void DiffViewer::setup_properties()
+{
+    set_font(Gfx::Font::default_fixed_width_font());
+    set_background_role(ColorRole::Base);
+    set_foreground_role(ColorRole::BaseText);
+}
+
+Vector<String> DiffViewer::split_to_lines(const String& text)
+{
+    // NOTE: This is slightly different thatn text.split('\n')
+    Vector<String> lines;
+    size_t next_line_start_index = 0;
+    for (size_t i = 0; i < text.length(); ++i) {
+        if (text[i] == '\n') {
+            auto line_text = text.substring(next_line_start_index, i - next_line_start_index);
+            lines.append(move(line_text));
+            next_line_start_index = i + 1;
+        }
+    }
+    lines.append(text.substring(next_line_start_index, text.length() - next_line_start_index));
+    return lines;
+}
+
+Gfx::Color DiffViewer::red_background()
+{
+    static Gfx::Color color = Gfx::Color::from_rgba(0x88ff0000);
+    return color;
+}
+
+Gfx::Color DiffViewer::green_background()
+{
+    static Gfx::Color color = Gfx::Color::from_rgba(0x8800ff00);
+    return color;
+}
+
+Gfx::Color DiffViewer::gray_background()
+{
+    static Gfx::Color color = Gfx::Color::from_rgba(0x88888888);
+    return color;
+}
+
+void DiffViewer::update_content_size()
+{
+    if (m_hunks.is_empty()) {
+        set_content_size({ 0, 0 });
+        return;
+    }
+
+    size_t num_lines = 0;
+    size_t current_original_line_index = 0;
+    for (const auto& hunk : m_hunks) {
+        num_lines += ((int)hunk.original_start_line - (int)current_original_line_index);
+
+        num_lines += hunk.removed_lines.size();
+        if (hunk.added_lines.size() - hunk.removed_lines.size()) {
+            num_lines += ((int)hunk.added_lines.size() - (int)hunk.removed_lines.size());
+        }
```

Comment:
Is this `if` necessary? `num_lines += (int)(hunk.added_lines.size() - hunk.removed_lines.size())`?

Reply:
oops, means to do:
```c++
if (hunk.added_lines.size() > hunk.removed_lines.size())
```